use_coverage causes "Can only merge Directories with no duplicates" error for Go packages with _test packaged files

**Describe the bug**
When `use_coverage` is enabled, Go packages that have `<package>_test` packaged test files (ie. for the package `foo`, the tests are in the same folder but are in the `foo_test` package and import `foo`, see [reproduction repository for an example](https://github.com/nickbruun/pants-use-coverage-repro/tree/main/gocode/pkg/repro)) causes an `IntrinsicError` of the following shape when the `test` goal is run:

    IntrinsicError: Can only merge Directories with no duplicates, but found 2 duplicate entries in __pkgs__/foo_pkg_repro:

    `__pkg__.a`: 1.) file digest=58ba11fe1afd8ddcccdd2a69c32f391ab4314b10430c13dfc8e00501f2368b5f size=1960:

    <snip>

    `__pkg__.a`: 2.) file digest=085a4521ca0c1cca011a41fc59dd11487677270df92568eead5cd16548eeb5da size=8798:

    <snip>

A simple reproduction case can be found at [nickbruun/pants-use-coverage-repro](https://github.com/nickbruun/pants-use-coverage-repro). The repository has two commits:

1. https://github.com/nickbruun/pants-use-coverage-repro/commit/c3ca0a0f715725cbe143a42451cf58b8248b7031: With `use_coverage` enabled. Failure can be seen in [GitHub Actions log](https://github.com/nickbruun/pants-use-coverage-repro/actions/runs/10766164255/job/29851474286)
2. https://github.com/nickbruun/pants-use-coverage-repro/commit/0b7a49d4a1f99668ebe881eb5bcd263a155ae647: With `use_coverage` disabled. The successful test run can be seen in [GitHub Actions log](https://github.com/nickbruun/pants-use-coverage-repro/actions/runs/10766186669/job/29851535601).

**Pants version**
2.21.1, but confirmed at least as far back as 2.17 (I just hadn't gotten around to reporting it until now -- sorry!)

**OS**
Reproduced on Linux and MacOS.

**Additional info**
See https://github.com/nickbruun/pants-use-coverage-repro/pull/1 for evidence that if the test has the non-`_test` package, the [test goal succeeds](https://github.com/nickbruun/pants-use-coverage-repro/pull/1/checks).

The following addition to `src/python/pants/backend/go/goals/test_test.py` will reproduce the issue in tests:

```py
def test_external_test_with_use_coverage(rule_runner: RuleRunner) -> None:
    rule_runner.write_files(
        {
            "foo/BUILD": "go_mod(name='mod')\ngo_package()",
            "foo/go.mod": "module foo",
            "foo/add.go": textwrap.dedent(
                """
                package foo
                func Add(x, y int) int {
                  return x + y
                }
                """
            ),
            "foo/add_test.go": textwrap.dedent(
                """
                package foo_test
                import (
                  "foo"
                  "testing"
                )
                func TestAdd(t *testing.T) {
                  if foo.Add(2, 3) != 5 {
                    t.Fail()
                  }
                }
                """
            ),
        }
    )
    tgt = rule_runner.get_target(Address("foo", generated_name="./"))
    rule_runner.set_options(
        [
            "--test-use-coverage",
        ],
        env_inherit={"PATH"},
    )
    result = rule_runner.request(
        TestResult, [GoTestRequest.Batch("", (GoTestFieldSet.create(tgt),), None)]
    )
    assert result.exit_code == 0
```

That code will not compile with the regular Go toolchain. `foo` and `foo_test` are both in the same package `foo` which cannot import itself. The Pants error with digests is confusing but the code will fail to compile at a later stage even if it got past the Pants digest errors.

The error with the Go toolchain is:

```
foo_test.go:4:2: package foo is not in std (/usr/local/Cellar/go/1.23.1/libexec/src/foo)
```

with the following files:

go.mod:
```
module pantsbuild.org/test-repro

go 1.23.1
```

foo.go:
```
package foo

import (
	"fmt"
)

func Grok() {
	fmt.Println("Hello World")
}
```

foo_test.go:
```
package foo

import (
	"foo"
	"testing"
)

func TestGrok(
	foo.Grok()
}
```

If you really want to import the `foo` package into the test package, then you need to use the _external test_ mechanism (that is, an `xtest` package). Declare your test as belonging to `package foo_xtest` instead of `package foo_test`.

--- a/docs/notes/2.24.x.md
+++ b/docs/notes/2.24.x.md
@@ -76,6 +76,10 @@ Pants will now warn if any errors are encountered while fingerprinting candidate
 
 Added a new `cache_scope` field to `adhoc_tool` and `shell_command` targets to allow configuration of the "cache scope" of the invoked process. The cache scope determines how long Pants will cache the result of the invoked process absent any other invalidation of the result via source or dependency changes.
 
+#### Go
+
+Fix a bug where Pants raised an internal exception which occurred when compiling a Go package with coverage support when the package also had an external test which imported the base package.
+
 ### Plugin API changes
 
 The `path_metadata_request` intrinsic rule can now access metadata for paths in the local system outside of the build root. Use the new `namespace` field on `PathMetadataRequest` to request metdata on local system paths using namespace `PathNamespace.SYSTEM`.

--- a/src/python/pants/backend/go/goals/test_test.py
+++ b/src/python/pants/backend/go/goals/test_test.py
@@ -714,3 +714,45 @@ def test_profile_options_write_results(rule_runner: RuleRunner) -> None:
         "test_runner",
         "trace.out",
     ]
+
+
+def test_external_test_with_use_coverage(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "foo/BUILD": "go_mod(name='mod')\ngo_package()",
+            "foo/go.mod": "module foo",
+            "foo/add.go": textwrap.dedent(
+                """
+                package foo
+                func Add(x, y int) int {
+                  return x + y
+                }
+                """
+            ),
+            "foo/add_test.go": textwrap.dedent(
+                """
+                package foo_test
+                import (
+                  "foo"
+                  "testing"
+                )
+                func TestAdd(t *testing.T) {
+                  if foo.Add(2, 3) != 5 {
+                    t.Fail()
+                  }
+                }
+                """
+            ),
+        }
+    )
+    tgt = rule_runner.get_target(Address("foo", generated_name="./"))
+    rule_runner.set_options(
+        [
+            "--test-use-coverage",
+        ],
+        env_inherit={"PATH"},
+    )
+    result = rule_runner.request(
+        TestResult, [GoTestRequest.Batch("", (GoTestFieldSet.create(tgt),), None)]
+    )
+    assert result.exit_code == 0

--- a/src/python/pants/backend/go/util_rules/build_pkg_target.py
+++ b/src/python/pants/backend/go/util_rules/build_pkg_target.py
@@ -458,7 +458,10 @@ async def setup_build_go_package_target_request(
         maybe_base_pkg_dep = await Get(
             FallibleBuildGoPackageRequest,
             BuildGoPackageTargetRequest(
-                request.address, for_tests=True, build_opts=request.build_opts
+                request.address,
+                for_tests=True,
+                with_coverage=request.with_coverage,
+                build_opts=request.build_opts,
             ),
         )
         if maybe_base_pkg_dep.request is None:

